src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -1665,7 +1807,12 @@ public int hashCode() {
             */
                 if (firstOnly) return ret;
             }
-            if (getHeader().getSequenceDictionary().size() == 0) {
+            if (getHeader() == null) {
+                if (ret == null) ret = new ArrayList<SAMValidationError>();
+                ret.add(new SAMValidationError(SAMValidationError.Type.MISSING_HEADER, "A non-null SAMHeader is required to resolve the sequence dictionary for a mapped read.", getReadName()));
+                if (firstOnly) return ret;
+            }
```

Comment:
Not sure I like this...the whole point of this PR was to make headerless `SAMRecords` an officially supported, valid state for a read to be in. Yet with this change they will fail `isValid()`. Can't we instead say that headerless `SAMRecords` are valid, but are validated to a less strict standard (ie., some checks that require a header are not performed)?